### PR TITLE
feat(appsync): add authenticationConfig to HttpDataSource, fixes #9934

### LINKED ISSUE #9934: [aws-appsync] Adding AuthorizationConfig within construction HttpDataSource of the appsync module.
Adding AuthorizationConfig within construction HttpDataSource of the appsync module.

### Use Case

Integrating AWS StepFunctions to the appsync as HTTP Datasource.
Current(24, August 2021) AWS CDK can not configure the authorizationConfig during creating data source.
Currently, That can do only from the AWS-CLI. (create-data-source command) 
https://docs.aws.amazon.com/cli/latest/reference/appsync/create-data-source.html

Related AWS Blogs.
https://aws.amazon.com/blogs/mobile/invoke-aws-services-directly-from-aws-appsync/

### Proposed Solution

I'd like to call the addHttpDataSource method with authorizationConfig object.

```javascript
    api.addHttpDataSource('ds', endpoint, {
      name: 'custom',
      description: 'custom description',
      authorizationConfig: {
        authorizationType: 'AWS_IAM',
        awsIamConfig: {
          signingRegion: 'us-east-1',
          signingServiceName: 'states'
        }
      }
    });
```

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by haruharuharuby**
### Description
Adding authorization config to HttpDatasource in aws-appsync module.
Users will be able to specify the awsIamConfig in httpConfig as parameter of HttpDataSource.

### Pull Request Checklist
* [x] Testing
  [x] Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  [-] __CLI change?:__ coordinate update of integration tests with team
  [-] __cdk-init template change?:__ coordinated update of integration tests with team

* [ ] Docs
  [-] __jsdocs__: All public APIs documented
  [-] __README__: README and/or documentation topic updated
  [-] __Design__: For significant features, design document added to `design` folder

* [ ] Title and Description
  [x-]  __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  [x] __Title__: use lower-case and doesn't end with a period
  [-] __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  [x] __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"

* [x] Sensitive Modules (requires 2 PR approvers)
  [-]  IAM Policy Document (in @aws-cdk/aws-iam)
  [-] EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  [-] Grant APIs (only if not based on official documentation with a reference)

Fixes: #9934

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appsync/lib/data-source.ts [hunk 1]
```diff
@@ -203,6 +203,35 @@ export class DynamoDbDataSource extends BackedDataSource {
...
+  /**
+   * The signing region for AWS IAM authorization
+   */
+  readonly signingRegion: string;
+
+  /**
+   * The signing service name for AWS IAM authorization
+   */
+  readonly signingServiceName: string;
+}
```
Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/lib/data-source.ts:233]: If the authorization type is always going to AWS_IAM as seen by these [docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appsync-datasource-authorizationconfig.html), I think it might be best to just expose `awsIamConfig` and remove that additional level of abstraction!

Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/lib/data-source.ts:233]:
```suggestion
export interface AwsIamConfig {
  /**
   * The signing region for AWS IAM authorization
   */
  readonly signingRegion: string;

  /**
   * The signing service name for AWS IAM authorization
   */
  readonly signingServiceName: string;
}
```

File: packages/@aws-cdk/aws-appsync/lib/data-source.ts [hunk 2]
```diff
@@ -211,6 +240,7 @@ export interface HttpDataSourceProps extends BaseDataSourceProps {
    * The http endpoint
    */
   readonly endpoint: string;
+  readonly authorizationConfig?: HttpDataSourceAuthorizationConfig;
```
Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/lib/data-source.ts:243]:
```suggestion
  readonly authorizationConfig?: AwsIamConfig; 
```

Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/lib/data-source.ts:243]: remove layer of abstraction and let CDK handle the heavy lifting :) repeat elsewhere

File: packages/@aws-cdk/aws-appsync/lib/data-source.ts (hunk 1 above)
Comment (haruharuharuby) [reply packages/@aws-cdk/aws-appsync/lib/data-source.ts]: @BryanPan342 I have a question.

JSII does not allow nested objects directly. I got a linting error when I tried to implement like below.

`Only string-indexed map types are supported`

```ts
export interface AwsIamConfig {
  /**
   * The authorization type required by the HTTP endpoint
   */
  readonly authorizationType: 'AWS_IAM';
  /**
   * The IAM configuration required by the HTTP endpoint
   */
  readonly awsIamConfig: {
    /**
     * The signing region for AWS IAM authorization
     */
    readonly signingRegion: string;

    /**
     * The signing service name for AWS IAM authorization
     */
    readonly signingServiceName: string;
  }
}
`

Do you know the better way to integrate it?

File: packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts [hunk 3]
```diff
@@ -57,6 +57,35 @@ describe('Http Data Source configuration', () => {
...
+    // WHEN
+    api.addHttpDataSource('ds', endpoint, {
+      name: 'custom',
+      description: 'custom description',
+      authorizationConfig: {
+        authorizationType: 'AWS_IAM',
+        awsIamConfig: {
+          signingRegion: 'us-east-1',
+          signingServiceName: 'states'
+        }
```
[the remark this answers is not in the document]
Comment (haruharuharuby) [reply packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts:70]: I was misunderstanding your comment. It would be better to configure without the authorizationType because the authorizationType takes a single value.  Is it matched what you wanted to point it out?

File: packages/@aws-cdk/aws-appsync/README.md [hunk 4]
```diff
@@ -18,8 +18,10 @@ APIs that use GraphQL.
 
 ### Example
 
+#### DynamoDB Data source example
```
Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/README.md:21]:
```suggestion
#### DynamoDB
```

File: packages/@aws-cdk/aws-appsync/README.md [hunk 5]
```diff
@@ -82,6 +84,87 @@ demoDS.createResolver({
 });
 ```
 
+#### Http data source example. Integrating SterFunctions workflow.
```
Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/README.md:87]:
```suggestion
#### HTTP Endpoints
```

File: packages/@aws-cdk/aws-appsync/README.md [hunk 6]
```diff
@@ -82,6 +84,88 @@ demoDS.createResolver({
...
+
+httpDs.createResolver({
+  typeName: 'Mutation',
+  fieldName: 'callStepFunction',
+  requestMappingTemplate: MappingTemplate.fromFile('request.vtl'),
+  responseMappingTemplate: MappingTemplate.fromFile('response.vtl')
+});
+```
+
+## Imports
```
[the remark this answers is not in the document]
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/README.md:167]: should be on line 213

File: packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (BryanPan342) [reply packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts:70]: oh my bad.. i apologize i definitely misread this it LGTM

**Comment by BryanPan342** [comment]
i believe the build is failing for this reason

```
error TS0: @aws-cdk/aws-appsync.GraphqlApiBase#addHttpDataSource changes type of argument options when implementing @aws-cdk/aws-appsync.IGraphqlApi (expected @aws-cdk/aws-appsync.DataSourceOptions, found @aws-cdk/aws-appsync.HttpDataSourceOptions
```

https://github.com/aws/aws-cdk/blob/beb02691e0187180b4042ead4562a66142cd6bdc/packages/%40aws-cdk/aws-appsync/lib/graphqlapi-base.ts#L80

this needs to be changed to `HttpDataSourceOptions` instead of `DataSourceOptions`

**Comment by haruharuharuby** [comment]
@BryanPan342 I've just pushed the code that removing nested construction. Thank you for reviewing.
https://github.com/aws/aws-cdk/pull/9971/commits/e6250b2f4f4bf45b4f398dd9dbb0c70550f81cf3

**Comment by haruharuharuby** [comment]
@BryanPan342 I want to solve the CodeBuild error. Can you show me how to look for the Error in the CodeBuild log?

**Comment by BryanPan342** [comment]
> i believe the build is failing for this reason
> 
> ```
> error TS0: @aws-cdk/aws-appsync.GraphqlApiBase#addHttpDataSource changes type of argument options when implementing @aws-cdk/aws-appsync.IGraphqlApi (expected @aws-cdk/aws-appsync.DataSourceOptions, found @aws-cdk/aws-appsync.HttpDataSourceOptions
> ```
> 
> https://github.com/aws/aws-cdk/blob/beb02691e0187180b4042ead4562a66142cd6bdc/packages/%40aws-cdk/aws-appsync/lib/graphqlapi-base.ts#L80
> 
> this needs to be changed to `HttpDataSourceOptions` instead of `DataSourceOptions`

@haruharuharuby this is the error + fix

if you hit the `build logs` link in the AWS CodeBuild CI Report then you will get directed to the output

then Ctrl-F for the `!!!!!` or scroll all the way to the bottom of the log.

**Comment by haruharuharuby** [comment]
#9934

File: packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts [hunk 7]
```diff
@@ -57,6 +57,33 @@ describe('Http Data Source configuration', () => {
     });
   });
 
+  test('appsync configures name, authorizationConfig correctly', () => {
+    // WHEN
+    api.addHttpDataSource('ds', endpoint, {
+      name: 'custom',
+      description: 'custom description',
+      authorizationConfig: {
+        authorizationType: 'AWS_IAM',
```
Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts:66]: why is this line necessary? the authorizationType is fixed to AWS_IAM no?

File: packages/@aws-cdk/aws-appsync/lib/data-source.ts [hunk 8]
```diff
@@ -203,6 +203,25 @@ export class DynamoDbDataSource extends BackedDataSource {
...
 }
 
+/**
+ * The authorization config in case the HTTP endpoint requires authorization
+ */
+export interface AwsIamConfig {
+  /**
+   * The authorization type required by the HTTP endpoint
+   */
+  readonly authorizationType: 'AWS_IAM';
```
Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/lib/data-source.ts:213]: is this field necessary if it's fixed?

could we just make the interface

```ts
export interface AwsIamConfig {
  readonly signingRegion: string;
  readonly signingServiceName: string;
}
```

File: packages/@aws-cdk/aws-appsync/lib/data-source.ts [hunk 9]
```diff
@@ -211,19 +230,38 @@ export interface HttpDataSourceProps extends BaseDataSourceProps {
...
+          authorizationConfig: {
+            authorizationType: authConfig.authorizationType,
+            awsIamConfig: {
+              signingRegion: authConfig.signingRegion,
+              signingServiceName: authConfig.signingServiceName
+            } 
+            
+          }
+        },
+        type: 'HTTP',
```
Comment (BryanPan342) [inline packages/@aws-cdk/aws-appsync/lib/data-source.ts:262]:
```suggestion
   const authorizationConfig = props.authorizationConfig ? {
     authorizationType: 'AWS_IAM',
     awsIamConfig: props.authorizationConfig,
   } : undefined;
   
   super(scope, id, props, {
     httpConfig: {
       endpoint: props.endpoint,
       authorizationConfig,
    },
  }); 
```

File: packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts (hunk 7 above)
Comment (haruharuharuby) [reply packages/@aws-cdk/aws-appsync/test/appsync-http.test.ts]: Yes, It true. AWS_IAM takes only a single value. I remove it.

File: packages/@aws-cdk/aws-appsync/lib/data-source.ts (hunk 8 above)
Comment (haruharuharuby) [reply packages/@aws-cdk/aws-appsync/lib/data-source.ts]: So the authorizationType removes from the caller.  Building it Constructor in DataSource.

**Comment by haruharuharuby** [comment]
@BryanPan342 I've just passed CI process. (fixed some linting and Unit test)

**Review by BryanPan342 — DISMISSED** [review]
Woooo LGTM 🎉

Thanks for the contribution this looks great!

I'll let @MrArnoldPalmer give his blessing before its shipped 😊

**Comment by haruharuharuby** [comment]
@BryanPan342 Thank you for a lot of reviews. :)

**Review by MrArnoldPalmer — DISMISSED** [review]

**Comment by BryanPan342** [comment]
@haruharuharuby I think you need to allow maintainers to edit your PR in order for Mergify to work

**Comment by haruharuharuby** [comment]
@BryanPan342 Sorry I made this PR from my org account. So suppose from this doc,  I can't edit permission. 
So I could not find that menu.

https://github.community/t/how-can-we-enable-allow-edits-from-maintainers-by-default/2847

is it necessary to remake PR?

**Comment by BryanPan342** [comment]
> @BryanPan342 Sorry I made this PR from my org account. So suppose from this doc, I can't edit permission.
> So I could not find that menu.
> 
> https://github.community/t/how-can-we-enable-allow-edits-from-maintainers-by-default/2847
> 
> is it necessary to remake PR on my regular account (haruharuharuby)

@haruharuharuby im guessing that's the right move? if you upload it with the same edits i can approve of it right away

**Comment by haruharuharuby** [comment]
@BryanPan342 Thank you for following up. I made another PR by @haruharuharuby the same as this PR.
Please check it.

https://github.com/aws/aws-cdk/pull/10171

**Comment by haruharuharuby** [comment]
> @haruharuharuby im guessing that's the right move? if you upload it with the same edits i can approve of it right away

Yes, I think the only way to mearging is to make another PR.  looking over in this PR(coming from Org account), I could not find any way that displaying the "allow changes by maintainers".

**Comment by BryanPan342** [comment]
Closing in favor of #10171
